sophia/src/serializer.rs

```diff
@@ -1,217 +1,355 @@
-//! Serializers for standard RDF syntaxes,
-//! and tools for building new serializers.
-//!
-//! # Uniform interface
-//!
-//! Each serializer module defines a `Config` type, that
-//! - implements [`Default`],
-//! - has a `writer` method taking any `io::Write`
-//!   and returning a [`TripleWriter`] or a [`QuadWriter`],
-//! - has a `stringifier` method returning a [`StringSerializer`] or a [`QuadStringifier`].
-//!
-//! Each serializer module also provides two functions `writer` and a `stringifier`,
-//! calling the corresponding methods from the default `Config`.
-//!
-//! [`TripleWriter`] and [`TripleStringifier`] are specializations of [`TripleSink`].
-//! [`QuadWriter`] and [`QuadStringifier`] are specializations of [`QuadSink`].
-//!
-//! [`Default`]: https://doc.rust-lang.org/std/default/trait.Default.html
-//! [`TripleWriter`]: trait.TripleWriter.html
-//! [`TripleStringifier`]: trait.TripleStringifier.html
-//! [`QuadWriter`]: trait.QuadWriter.html
-//! [`QuadStringifier`]: trait.QuadStringifier.html
-//! [`TripleSink`]: ../triple/stream/trait.TripleSink.html
-//! [`QuadSink`]: ../quad/stream/trait.QuadSink.html
-
-use std::convert::Infallible;
+//! API for serializing RDF syntaxes.
+
 use std::io;
 
 use crate::dataset::*;
 use crate::graph::*;
-use crate::quad::{stream::*, *};
+use crate::quad::stream::*;
 use crate::triple::{stream::*, *};
 
 use std::result::Result; // override ::error::Result
 
-#[macro_use]
-pub mod common;
-pub mod nq;
-pub mod nt;
-
-/// An extension of the `TripleSink` trait,
-/// dedicated to serialization to IO streams.
+/// A triple serializer formats triples in some target of type `T`.
+/// The triples may come from a [`TripleSource`] or a [`Graph`].
 ///
-/// See also [`TripleSink`].
-///
-/// [`TripleSink`]: ../triple/stream/trait.TripleSink.html
-///
-pub trait TripleWriter<W: io::Write>: TripleSink<Outcome = ()> + Sized {
-    type Config;
+/// [`TripleSource`]: ../triple/stream/trait.TripleSource.html
+/// [`Graph`]: ../graph/trait.Graph.html
+pub trait TripleSerializer<T> {
+    type Sink: TripleSink;
 
-    fn new(write: W, config: Self::Config) -> Self;
+    /// A sink serializing into `target` the triples fed to it.
+    fn sink(&self, target: T) -> Self::Sink;
```

Comment:
I think this architecture blurrs the lines what a serializer is. On the one hand there is `TripleSerializer` that offers methods to serialize (`serialize_triples_in(to)*()`) but it doesn't implement the serializing. On the other hand a sink must be implemented that actually does serialization but don't implements `TripleSerializer` ... 🤔 . Even worse, thinking about Turtle-serialization for streams and graphs it is required to implement two different algorithms, i.e. different 'sinks' but the trait only allows one associated sink.

I suggest a more plain trait the does not include construction but focuses more on serialization. Indeed, my proposal is pretty close to the original and is similar to a `TripleSink`.

```rust
trait TripleSerializer {
    type Outcome;
    type Error: std::error::Error;

    fn serialize_term<TD: TermData>(&mut self, t: &Term<TD>) -> Result<(), Self::Error>;

    fn serialize_triple<T: Triple>(&mut self, t: &T) -> Result<(), Self::Error>;

    fn serialize_triples<TS: TripleSource>(&mut self, ts: TS) -> Result<(), StreamError<TS::Error, Self::Error>>;

    fn serialize_graph<G>(&mut self, g: &G) -> Result<(), StreamError<G::Error, Self::Error>>;

    fn finish(self) -> Result<Self::Outcome, Self::Error>;
}
```

A `TripleStringifier` is would be similar just return `String` instead of `()` and provide a blanked implementation for `TripleSerializer<Outcome=Vec<u8>>`. However, this proposal probably needs some refinement.

---

\* in this case the preposition 'in' indicates me that the serialization is done somewhere inside without modifying the outer thing. I'd rather go with 'into'. For me this is a more common description in Rust for: Take something convert it and modify the target. (BTW the same for `in_sink()`.)

Reply:
Here are the two design principles that I tried to follow when refactoring the parser and serializer APIS:

1. be intuitive: a Parser should have a `parse` method (or several); a Serializer should have a `serialize` method (or several);
2. integrate well with the general streaming model of `sophia`, i.e. provide implementations of `TripleSource`/`TripleSink` (or their quad counterpart) whenever possible.

You object that the actual serializing is done in the sink and not in the serializer *per se*, but the same criticism could be done on parsers as well (the actual parsing is done by the source returned by the `parse` method).

Note also that I have been aware of this "inconsistency" from the start. This very concern lead me to the previous design, naming the "static" object with parse/serialize methods "Config" rather than "Parser"/"Serializer". This was deemed unintuitive by several people (including yourself, IIRC :wink:), hence the new design (and item 1 above).

> Even worse, thinking about Turtle-serialization for streams and graphs it is required to implement two different algorithms,

Serializing a stream and serializing (smartly) a graph do indeed require two different algorithms. So it might be desirable to have two different traits... On the down side, this would make the API more complex. On the up side, you might say that implementors could choose to implement only the graph serialization, and *not* the stream serialization (the opposite is not a problem, given the default implementation of  `serialize_graph_in`). But my feeling is that once you have a the algorithm for graphs, the algorithm for streams is really not that hard to get...

>  i.e. different 'sinks'

No, the sink is only for parsing streams.

Reply:
Re. the `TripleSerializer` trait you proposed:

* `serialize_term` does not make sense for some serializers (i.e. RDF/XML), where terms can not always be "isolated";

* `serialize_triple` is not strictly required by users (they can easily reuse `serizlier_triples` or `serialize_graph` to get the same result); and for internal use, such a method does not always make sense (in Turtle or RDF/XML, several triples share the same subject, so not all triples are serialized identically);

* `serializer_triples` means that your design also suffers from the "two distinct algorithms" problem discussed above.

Reply:
Re. "in" vs. "into":

I have given this some thought when designing the stream API – first I wanted to name the methods `into_sink` and `into_graph` of `TripleSource`, but this was misleading: it gave the impression that the methods were for *converting* the `TripleSource` into something else, which is wrong.

So I went for  `in` instead of `into`, and I kept this habit for serializers.

Reply:
> Note also that I have been aware of this "inconsistency" from the start. This very concern lead me to the previous design, naming the "static" object with parse/serialize methods "Config" rather than "Parser"/"Serializer". This was deemed unintuitive by several people (including yourself, IIRC 😉), hence the new design (and item 1 above).

I have to admit that this design is actually pretty good. And the same I come up with on a second thought. However, I still believe that the initialization should not be part of a basic trait.

Indeed, the methods `serialize_term()` and `serialize_triple()` add no benefits and can be removed.

Regarding the 'in' vs 'into' discussion... why not name it `serialize_triples()` and `serialize_graph()`? This does exactly describe what is done and the API clearly tells the user if the argument is consumed (`TripleSource`) or just borrowed (`Graph`). For me this is one of the biggest advantages of Rust's typesystem 👍 .

To sum up, I now propose to stick with my previous proposed trait but remove the two unnecessary methods. Maybe add `: TripleSink` to get `Self::Outcome` and `Self::Error`.

Reply:
> However, I still believe that the initialization should not be part of a basic trait.

I agree with this general principle. But it seems that we are placing the "initialization" at different levels.

I'm assuming that by "initialization", you are referring to setting the *target* (a file in general) of the serializer. In my view, this is already a second step; the first step is setting the different options of the serialier, like "should it use ASCII or UTF-8" (for the N-Triples serializer) or "what prefixes should it use" (for a future Turtle serializer)...

I think those two steps should be separate. I think it is important that one can configure, say, a turtle serializer with a number of option (typically, which prefixes it should use), and then use that serializer several times, to serialize different graphs in different files...

Also, I'm trying to make the Parser and Serializer APIs as symmetrical as possible. But I give you that the symmetry is not perfect in the current state. I'll come with another proposal, in the main thread of this PR.